### PR TITLE
Modify distributed transaction commit flow

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR updates the commit flow as per the new design doc on atomic distributed transactions.
- Any failure before a commit decision rollback the transaction.
- Any failure before conclude, records a warning and return an error.
- The user can track the transaction state by issuing `show warnings` and then extracting transaction ID. Later can issue `show transaction status for '<id>'` to know the current state of it.

## Related Issue(s)

- #16245 

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by GuptaManan100 — APPROVED** [review]
Logic LGTM

File: go/vt/vtgate/debug_2pc.go [hunk 1]
```diff
@@ -18,4 +18,52 @@ limitations under the License.
...
+	"vitess.io/vitess/go/vt/log"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
 const DebugTwoPc = true
+
+// checkTestFailure is used to simulate failures in 2PC flow for testing when DebugTwoPc is true.
+func checkTestFailure(ctx context.Context, expectCaller string, target *querypb.Target) error {
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/debug_2pc.go:34]: There are multiple reasons
1. This is an unused function in the production build. 
2. Anyone can end up using it if this implementation remains in production code.
3. Simple no-op production code will avoid any incidental calls to it.

File: go/vt/vtgate/tx_conn.go [hunk 2]
```diff
@@ -190,114 +197,118 @@ func (txc *TxConn) commit2PC(ctx context.Context, session *SafeSession) error {
...
 		return txc.tabletGateway.CommitPrepared(ctx, s.Target, dtid)
-	})
-	if err != nil {
+	}
+	if err = txc.runSessions(ctx, rmShards, session.logging, prepareCommitAction); err != nil {
 		return err
 	}
 
-	return txc.tabletGateway.ConcludeTransaction(ctx, mmShard.Target, dtid)
+	// At this point, application can continue forward.
```
Comment (systay) [inline go/vt/vtgate/tx_conn.go:271]: for my own understanding - could we in theory do the clean up in a background thread and answer the user that their commit has been accepted?

File: go/vt/vtgate/tx_conn.go [hunk 3]
```diff
@@ -62,6 +59,16 @@ var txAccessModeToEOTxAccessMode = map[sqlparser.TxAccessMode]querypb.ExecuteOpt
 	sqlparser.ReadOnly:               querypb.ExecuteOptions_READ_ONLY,
 }
 
+type commitPhase int
+
+const (
+	COMMIT2PC_CREATETRANSACTION = iota
```
Comment (systay) [inline go/vt/vtgate/tx_conn.go:65]: this looks weird. do you really want iota on the three first values?
and, the naming is not camel case which I believe is the golang standard

File: go/vt/vtgate/tx_conn.go [hunk 4]
```diff
@@ -62,6 +59,16 @@ var txAccessModeToEOTxAccessMode = map[sqlparser.TxAccessMode]querypb.ExecuteOpt
...
+	COMMIT2PC_PREPARE           = iota
+	COMMIT2PC_STARTCOMMIT       = iota
+	COMMIT2PC_PREPARECOMMIT
+	COMMIT2PC_CONCLUDE
+)
```
Comment (systay) [inline go/vt/vtgate/tx_conn.go:70]: You probably meant to use the type, and `iota` should generally only be used on the first value.

```suggestion
const (
	Commit2pcCreateTransaction commitPhase = iota
	Commit2pcPrepare
	Commit2pcStartCommit
	Commit2pcPrepareCommit
	Commit2pcConclude
)```

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/tx_conn.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/tx_conn.go:271]: In theory, yes, but it is better to call conclude now, than using goroutine or transaction resolver to conclude.

**Pull request merged**